pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+        fn build(&self) {
+            assert!(self.params.is_valid());
+
+            let now = frame_system::Pallet::<T>::block_number();
+            let (max_emission, config) = Pallet::<T>::recalculate_inflation(now);
+
+            ActiveInflationConfig::<T>::put(config);
+            SafetyInflationTracker::<T>::put(InflationTracker {
+                cap: max_emission,
+                issued: 0,
+            });
+            InflationParams::<T>::put(self.params);
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumber> for Pallet<T> {
+        fn on_initialize(now: BlockNumber) -> Weight {
+            let recalculation_weight =
+                if Self::is_recalculation_in_next_block(now, &ActiveInflationConfig::<T>::get()) {
+                    T::WeightInfo::hook_with_recalculation()
+                } else {
+                    T::WeightInfo::hook_without_recalculation()
+                };
+
+            // Benchmarks won't acount for whitelisted storage access so this needs to be added manually.
+            //
+            // SafetyInflationTracker - 1 DB read & write
+            // ActiveInflationConfig - 1 DB read
+            let whitelisted_weight =
+                <T as frame_system::Config>::DbWeight::get().reads_writes(2, 1);
+
+            recalculation_weight
+                .saturating_add(T::WeightInfo::on_timestamp_set())
+                .saturating_add(whitelisted_weight)
+        }
+
+        fn on_finalize(now: BlockNumber) {
+            // Recalculation is done at the block right before the re-calculation is supposed to happen.
+            // This is to ensure all the rewards are paid out according to the new inflation configuration from next block.
+            //
+            // If this was done in `on_initialize`, collator & treasury would receive incorrect rewards for that one block.
+            //
+            // This should be done as late as possible, to ensure all operations that modify issuance are done.
+            if Self::is_recalculation_in_next_block(now, &ActiveInflationConfig::<T>::get()) {
+                let (max_emission, config) = Self::recalculate_inflation(now);
+                ActiveInflationConfig::<T>::put(config.clone());
+
+                SafetyInflationTracker::<T>::mutate(|tracker| {
+                    tracker.cap.saturating_accrue(max_emission);
+                });
+
+                Self::deposit_event(Event::<T>::NewInflationConfiguration { config });
+            }
+        }
+
+        fn integrity_test() {
+            assert!(T::CycleConfiguration::periods_per_cycle() > 0);
+            assert!(T::CycleConfiguration::eras_per_voting_subperiod() > 0);
+            assert!(T::CycleConfiguration::eras_per_build_and_earn_subperiod() > 0);
+            assert!(T::CycleConfiguration::blocks_per_era() > 0);
+        }
+    }
+
+    impl<Moment, T: Config> OnTimestampSet<Moment> for Pallet<T> {
```

Comment:
Dumb question: why payout is issued in `OnTimestampSet` hook?

Reply:
Far from dumb question - I just adopted the approach from before.
But I guess it's no longer needed, we could do the payout in `on_initialize` as well.